examples/features/torch_batch_process/batch_inference/compare_with_core_api/core_api_inference.py

```diff
@@ -0,0 +1,151 @@
+    # do not run into error
+    pathlib.Path.mkdir(pathlib.Path(inference_output_dir), parents=True, exist_ok=True)
+
+    dataloader_iterator = iter(data_loader)
+
+    with torch.no_grad():
+        for batch_idx in range(skip, per_worker_iterate_length):
+            X = next(dataloader_iterator, None)
+            logging.info(f"Working on batch is {batch_idx}")
+            if X is not None:
+                data, label = X
+                output = model(data)
+                preds = output.argmax(dim=1, keepdim=True)
+
+                file_name = f"inference_out_{rank}_{batch_idx}.json"
+
+                output = []
+
+                for pred in preds:
+                    output.append(pred[0].item())
+
+                with open(
+                    os.path.join(inference_output_dir, f"{file_name}"),
+                    "w",
+                ) as f:
+                    json.dump({"predictions": output}, f)
+
+                # After each batch, synchronize and update number of catches completed
+                if context.distributed.rank == 0:
+                    work_completed_this_round = sum(context.distributed.gather(len(data)))
+                    records_processed += work_completed_this_round
+                    checkpoint_metadata = {
+                        "steps_completed": batch_idx,
+                    }
+                    with context.checkpoint.store_path(checkpoint_metadata) as (path, uuid):
+                        with open(os.path.join(path, "batch_completed.json"), "w") as file_obj:
+                            json.dump({"batch_completed": batch_idx}, file_obj)
+                else:
+                    context.distributed.gather(len(data))
+
+                if context.preempt.should_preempt():
+                    return
+
+
+def _get_data_loader(batch_size: int, total_worker: int, rank: int, skip: int) -> [Any, int]:
+    transform = transforms.Compose(
+        [transforms.ToTensor(), transforms.Normalize((0.5, 0.5, 0.5), (0.5, 0.5, 0.5))]
+    )
+    with filelock.FileLock(os.path.join("/tmp", "inference.lock")):
+        inference_data = tv.datasets.CIFAR10(
+            root="/data", train=False, download=True, transform=transform
+        )
+    dataloader = pytorch.DataLoader(
+        dataset=inference_data,
+        batch_size=batch_size,
+        shuffle=False,
+    ).get_data_loader(repeat=False, skip=skip, num_replicas=total_worker, rank=rank)
+
+    # Enumerate over dataloader directly may cause some workers to iterate for 1 more time
+    # than others when drop_last = False. If those workers synchronize on the last batch_idx,
+    # they would hang forever as other workers never hit that last batch_idx.
+    # To avoid the issue, we calculate and take the ceiling of the iteration count to ensure
+    # all workers iterate for the same number of times.
+    per_worker_iterate_length = math.ceil(len(inference_data) / batch_size / total_worker)
```

Comment:
This is one way to do it, but if I were a user I would not be very happy with drop_last=False behavior; I'd really want to get predictions on every batch, exactly once.

Reply:
Agree that we want predictions on every batch, exactly once behavior, which is the current behavior of this example. `per_worker_iterate_length` simply guarantees each worker goes through the same number of step, the way we iterate through the dataloader guarantees no duplicated work is done.

`per_worker_iterate_length` is used in:
```
dataloader_iterator = iter(data_loader)       
for batch_idx in range(skip, per_worker_iterate_length):
            X = next(dataloader_iterator, None) 
            if X is not None:
                # run inference
            steps_completed = batch_idx + 1
            if steps_completed % checkpoint_interval == 0:
               # sync and checkpoint 
```
The reason why we do not simply enumerate dataloader is that, it may lead to the case where some workers iterate for one more step than others, if that step happen to be a sync & checkpoint step, the example would hang.

## An example
Concrete, imagine we have a dataset with 24 items, batch size = 10, and num_workers = 2, and we checkpoint after every batch.

### If we simply enumerate through the dataloader

batch 0:
worker 1 processes item idx 0 - 9
worker 2 processes items idx  10 - 19
worker 1 and worker 2 sync, then worker 1 checkpoint

batch 1:
worker 1 processes item idx 20 - 23
worker 1 hangs waiting for worker 2 to sync but worker 2 already finished enumeration after previous batch.

### If we use per_worker_iterate_length
We guarantee the number of iteration is the same per worker. 

During batch 1, worker 2 simply go straight to sync and checkpoint part of the code while worker 1 would first finish processing items (idx: 20-23) then go to sync and checkpoint.

## Related comments
https://github.com/determined-ai/determined/pull/7274/files#r1258964292

Based on the above reasons, we cannot simply break when X is None. Also, we cannot use enumerate directly.

Reply:
gotcha.  I just misunderstood what I was reading (both code and comment).

Maybe there is a way to communicate the intention of the code a little clearer?

Maybe with names and comments like this:

```python
# Keep track of how many iterations the worker with the longest shard will need:
max_shard_length = math.ceil(len(inference_data) / batch_size / total_worker)
dataloader_iterator = iter(data_loader)       
for batch_idx in range(skip, max_shard_length):
            X = next(dataloader_iterator, None)
            if X is None:
                # our shard is done, but the longest shard is still going
                pass
            else:
                ...
            steps_completed = batch_idx + 1
            if steps_completed % checkpoint_interval == 0:
               # sync and checkpoint 
```

Eh, it's still not great.  You can take it or leave it.

I suppose you could also do the `max_shard_length` calculation like with an allgather.  That might be a little more obvious to the reader, without any measurable performance impact:
```python
max_shard_length = max(core_context.distributed.allgather(len(dataloader_iterator)))
```